ftdc/custom_format.go

```diff
@@ -152,77 +241,66 @@ func getFieldsForItem(item any) []string {
+	for name, stats := range data {
+		mapOrder = append(mapOrder, name)
+		fieldsForItem, err := getFieldsForStruct(reflect.TypeOf(stats))
+		if err != nil {
+			return nil, &schemaError{name, err}
+		}
+
+		for _, field := range fieldsForItem {
 			// We insert a `.` into every metric/field name we get a recording for. This property is
 			// assumed elsewhere.
-			fields = append(fields, fmt.Sprintf("%v.%v", key, field))
+			fields = append(fields, fmt.Sprintf("%v.%v", name, field))
 		}
 	}
 
 	return &schema{
 		mapOrder:   mapOrder,
 		fieldOrder: fields,
-	}
+	}, nil
 }
 
 // flatten takes an input `Datum` and a `mapOrder` from the current `Schema` and returns a list of
 // `float32`s representing the readings. Similar to `getFieldsForItem`, there are constraints on
 // input data shape that this code currently does not validate.
-func flatten(datum datum, schema *schema) []float32 {
+func flatten(datum datum, schema *schema) ([]float32, error) {
 	ret := make([]float32, 0, len(schema.fieldOrder))
 
 	for _, key := range schema.mapOrder {
 		// Walk over the datum in `mapOrder` to ensure we gather values in the order consistent with
 		// the current schema.
 		stats, exists := datum.Data[key]
 		if !exists {
-			logging.Global().Warn("Missing data. Need to know how much to skip in the output float32")
-			return nil
-		}
-
-		rVal := reflect.ValueOf(stats)
-		if rVal.Kind() == reflect.Pointer {
-			rVal = rVal.Elem()
+			//nolint
+			return nil, fmt.Errorf("Missing statser name. Name: %v", key)
 		}
 
-		// Use reflection to walk the member fields of an individual set of metric readings. We rely
-		// on reflection always walking fields in the same order.
-		//
-		// Note, while reflection in this way isn't intrinsically expensive, it is making more small
-		// function calls and allocations than some more raw alternatives. For example, we can have
-		// the "schema" keep a (field, offset, type) index and we instead access get a single unsafe
-		// pointer to each structure and walk out index to pull out the relevant numbers.
-		for memberIdx := 0; memberIdx < rVal.NumField(); memberIdx++ {
-			rField := rVal.Field(memberIdx)
-			switch {
-			case rField.CanInt():
-				ret = append(ret, float32(rField.Int()))
-			case rField.CanFloat():
-				ret = append(ret, float32(rField.Float()))
-			default:
-				// Embedded structs? Just grab a global logger for now. A second pass will better
-				// validate inputs/remove limitations. And thread through a proper logger if still
-				// necessary.
-				logging.Global().Warn("Bad number type. Type:", rField.Type())
-				// Ignore via writing a 0 and continue.
-				ret = append(ret, 0)
-			}
+		numbers, err := flattenStruct(reflect.ValueOf(stats))
+		if err != nil {
+			return nil, err
 		}
+		ret = append(ret, numbers...)
 	}
 
-	return ret
+	return ret, nil
+}
+
+func parse(rawReader io.Reader) ([]datum, error) {
+	logger := logging.NewLogger("")
+	logger.SetLevel(logging.ERROR)
```

Comment:
Just a note-to-self hoping that this logger eventually gets threaded into the larger logger registry so it's controllable via `log`.

Reply:
As a reminder, this isn't used by the viam-server. Just by some CLI tool

Reply:
Oh whoops yep; sometimes I get confused on what the reading and writing sides are.